edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.DiagnosticsComponent/Scraper.cs

```diff
@@ -0,0 +1,91 @@
+        readonly HttpClient httpClient;
+        readonly Lazy<IDictionary<string, string>> endpoints;
+        readonly ILogger logger;
+
+        public Scraper(IList<string> endpoints, ILogger logger)
+        {
+            this.httpClient = new HttpClient();
+            this.endpoints = new Lazy<IDictionary<string, string>>(() => endpoints.ToDictionary(e => e, this.GetUriWithIpAddress));
+            this.logger = logger;
+        }
+
+        string GetUriWithIpAddress(string endpoint)
+        {
+            this.logger.LogInformation($"Getting uri with Ip for {endpoint}");
+            Match match = UrlRegex.Match(endpoint);
+            if (!match.Success)
+            {
+                throw new InvalidOperationException($"Endpoint {endpoint} is not a valid URL in the format <protocol>://<host>:<port>/<parameters>");
+            }
+
+            var hostGroup = match.Groups["host"];
+            string host = hostGroup.Value;
+            var ipHostEntry = Dns.GetHostEntry(host);
+            var ipAddr = ipHostEntry.AddressList.Length > 0 ? ipHostEntry.AddressList[0].ToString() : string.Empty;
+            var builder = new UriBuilder(endpoint);
+            builder.Host = ipAddr;
+            string endpointWithIp = builder.Uri.ToString();
+            this.logger.LogInformation($"Endpoint = {endpoint}, IP Addr = {ipAddr}, Endpoint with Ip = {endpointWithIp}");
+            return endpointWithIp;
+        }
+
+        public void Dispose()
+        {
+            this.httpClient.Dispose();
+        }
+
+        public async Task<IDictionary<string, string>> ScrapeAsync(CancellationToken cancellationToken)
+        {
+            var metrics = new Dictionary<string, string>();
+            foreach (KeyValuePair<string, string> endpoint in this.endpoints.Value)
+            {
+                this.logger.LogInformation($"Scraping endpoint {endpoint.Key}");
+                string metricsData = await this.ScrapeEndpoint(endpoint.Value, cancellationToken);
+                metrics.Add(endpoint.Key, metricsData);
+            }
+
+            return metrics;
+        }
+
+        private async Task<string> ScrapeEndpoint(string endpoint, CancellationToken cancellationToken)
+        {
+            try
+            {
+                var result = await this.httpClient.GetAsync(endpoint, cancellationToken);
+                if (result.IsSuccessStatusCode)
+                {
+                    return await result.Content.ReadAsStringAsync();
+                }
+                else
+                {
+                    this.logger.LogInformation($"Error connecting to {endpoint}\nResult error code {result.StatusCode}");
+                    throw new InvalidOperationException($"Error connecting to {endpoint}");
+                }
+            }
+            catch (Exception e)
+            {
+                this.logger.LogInformation($"Error scraping endpoint {endpoint} - {e.Message}");
+                throw;
```

Comment:
where do you expect this exception will be handled? or it will throw up to the top and crash edgeAgent?

Reply:
It wouldn't have crashed the main program, but it shouldn't throw. I'll make it return empty string instead.